extensions/cache/runtime/src/main/java/io/quarkus/cache/Cache.java

```diff
@@ -0,0 +1,56 @@
+package io.quarkus.cache;
+
+import java.util.concurrent.Callable;
+
+import io.quarkus.cache.runtime.caffeine.CaffeineCache;
+import io.smallrye.mutiny.Uni;
+
+/**
+ * This interface can be used to interact with a cache programmatically and store, retrieve or delete cache values.
+ * Most of its operations are reactive.
+ */
+public interface Cache {
+
+    /**
+     * Returns the unique and immutable default key for the current cache. This key is used by the annotations caching API when
+     * a no-args method annotated with {@link CacheResult} or {@link CacheInvalidate} is invoked. It can also be used with the
+     * programmatic caching API.
+     * 
+     * @return default cache key
+     */
+    Object getDefaultKey();
+
+    /**
+     * Returns a lazy asynchronous action that will emit the cache value identified by {@code key}, obtaining that value from
+     * {@code valueLoader} if necessary.
+     * 
+     * @param <T> cache value type
+     * @param key cache key
+     * @param valueLoader value loader
+     * @return a lazy asynchronous action that will emit a cache value
+     * @throws NullPointerException if the key is {@code null}
+     */
+    <T> Uni<T> get(Object key, Callable<T> valueLoader);
```

Comment:
Hmmm, maybe `Callable` makes it easier to deal with APIs that throw checked exceptions?

Reply:
Either requires an allocation per invocation, e.g. a capturing lambda. This can be avoided by using `Function<K, V>` instead, like Caffiene's and `computeIfAbsent`.

Reply:
I agree the Javadoc needs to mention how exceptions will be propagated.

Before I do any change regarding the callable/supplier/lambda choice, I'm wondering what kind of exception should be thrown from the cache when a computation fails. Right now, a `RuntimeException` is thrown. Should we have a dedicated exception to allow a catch from the user? `CacheException extends RuntimeException` maybe?

Reply:
@ben-manes has a point there. Supplier/Callable forces a capturing lambda. `Function<Object, T>` would be better here. 

Btw I see cache key type is not using generics. There are benefits to doing this. Any specific reason? One benefit in this particular case is that the Object that goes into `Function` could be the key, or the key + some other info. This makes it easier to make pure functions that do not capture external state.

Reply:
@gwenneg Any exception should be runtime. A `CacheException extends RuntimeException` would be good I think. It would allow the caller to different/treat cache specific issues differently easily.

Reply:
I replaced `Callable` with `Function<K, V>` as suggested here. I tried to take advantage of generics in this new version. @galderz: Could you please review `Cache` and `CaffeineCache` and tell me if that's what you had in mind?

Reply:
It would seem nicer to obtain a `Cache<K, V>` and have it typed throughout. You might include some holes, like `Map.containsKey(Object)`, where being loose is idiomatic as it has no usage side effects. This is why you'll see those methods, like `invalidate(Object)` in Caffeine/Guava but annotated for type checking. Or you might close the holes as an idiosyncrasy of the Collections Framework and be more restrictive.

I am unsure how you will reconcile programmatic and annotation usage of the same cache. JCache is completely broken in annotation-based model as keys became unusable by listeners, writers, etc. Spring Cache does not suffer from this because it offers only the annotation-based model, so the vendor API must be used directly for more advanced cases and not intermixed. You might have to dictate that the key is your wrapper type if you expect both models to be usable at once.

Reply:
I thought about `Cache<K, V>` before my last commit, but it didn't look right considering a single cache can be used to store many differents objects types using the caching annotations.

Reply:
Good points made @ben-manes and @gwenneg. I'd probably keep the API as is, but as Ben said, we have to verify that interoperability works between annotation and API approach, before deciding on the final look of the API. IOW, we don't want close or make things impossible with an API that doesn't allow for interoperability.